stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,176 @@
+    """
+    Implementation of the GraphWave structural embedding algorithm from the paper:
+        "Learning Structural Node Embeddings via Diffusion Wavelets" (https://arxiv.org/pdf/1710.10321.pdf)
+
+    This class is minimally initialized with a StellarGraph object. Calling the flow function will return a tensorflow
+    DataSet that contains the GraphWave embeddings.
+    """
+
+    def __init__(self, G, scales="auto", num_scales=3, num_eigenvecs=-1):
+        """
+        Args:
+            G (StellarGraph): the StellarGraph object.
+            scales (str or list of floats): the wavelet scales to use. "auto" will cause the scale values to be
+                automatically calculated.
+            num_scales (int): the number of scales when scales = "auto".
+            num_eigenvecs (int): the number of eigenvectors to use. When set to `-1` the maximum number of eigenvectors
+                is calculated.
+        """
+
+        if not isinstance(G, StellarGraph):
+            raise TypeError("G must be a StellarGraph object.")
+
+        node_types = list(G.node_types)
+        if len(node_types) > 1:
+            raise TypeError(
+                "{}: node generator requires graph with single node type; "
+                "a graph with multiple node types is passed. Stopping.".format(
+                    type(self).__name__
+                )
+            )
+
+        # Create sparse adjacency matrix:
+        # Use the node orderings the same as in the graph features
+        self.node_list = G.nodes_of_type(node_types[0])
+        self.Aadj = G.to_adjacency_matrix(self.node_list)
+
+        # Function to map node IDs to indices for quicker node index lookups
+        # TODO: Move this to the graph class
+        node_index_dict = dict(zip(self.node_list, range(len(self.node_list))))
+        self._node_lookup = np.vectorize(node_index_dict.get, otypes=[np.int64])
+
+        adj = G.to_adjacency_matrix().tocoo()
+        degree_mat = diags(np.array(adj.sum(1)).flatten())
+        laplacian = degree_mat - adj
+
+        if num_eigenvecs == -1:
+            num_eigenvecs = laplacian.shape[0] - 2
+
+        self.eigen_vals, self.eigen_vecs = eigs(laplacian, k=num_eigenvecs)
+        self.eigen_vals = np.real(self.eigen_vals).astype(np.float32)
+        self.eigen_vecs = np.real(self.eigen_vecs).astype(np.float32)
+
+        if scales == "auto":
+
+            e2 = self.eigen_vals[self.eigen_vals > 0].min()
+            eN = self.eigen_vals.max()
+
+            min_scale = -np.log(0.95) / np.sqrt(eN * e2)
+            max_scale = -np.log(0.85) / np.sqrt(eN * e2)
+
+            scales = np.linspace(min_scale, max_scale, num_scales)
+
+        self.scales = scales
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
```

Comment:
hmm that's a good point, but I'm not sure what a decent alternative default would be. Maybe `0.1 * N` eigenvecs or `c * log(N)` (for some constant). thoughts?

Reply:
In the sense that its really graph dependant and will probably have to be found by trial in error yeah, its like a hidden layer size.

I guess we're trying to find a `k` such that `sum(exp(-s * e_i) ^ 2 for i in range(k))` approximately equals `sum(exp(-s * e_i) ^ 2 for all eigenvalues)`. But I don't know of any way decent way to do this. 

We could try increasing values of `k`, and terminate when the  `sum(exp(-s * e_k)` is small relative to `sum(exp(-s * e_i) ^ 2 for i in range(k))`. but we'd have to restart the eigenvalue process entirely for each `k`. and we can't factor out the found eiegenvectors either because that would make the matrix less sparse :(

Reply:
- I observed the same phenomenon for Watch Your Step and also with the reddit dataset, it appears that this is a feature of real world graphs. This makes the Chebyshev approach more attractive
- thats by design - GraphWave filters out the large eigenvalues with the low-pass filter `exp(-s)` 
- finding all eigenvalues is `N^3` regardless of sparsity and ins't a viable option
- I'll look into whether this explosion of `s` makes sense in the context of handling zero/ near zero eigenvalues
- good point! we can compute the largest eigenvalue and smallest eigenvalue pretty quickly so I'll add that in

Reply:
> thats by design - GraphWave filters out the large eigenvalues with the low-pass filter exp(-s)

Indeed, I was only pointing it out to tie into the other comment on `which="SM"`, and the concerns about noisy tiny eigenvalues.

> finding all eigenvalues is N^3 regardless of sparsity and ins't a viable option

Indeed, and I imagine computing the default `N - 2` eigenvalues is a similar complexity. I take so you're actively investigating alternatives?

Reply:
> Indeed, and I imagine computing the default N - 2 eigenvalues is a similar complexity. I take so you're actively investigating alternatives?

I now a fast method of searching the eigenvalues; 0.25s to recover 80% of the filter eigenvalue norm on Cora with `min_scale=4`. 

> Can we accurately compute the smallest eigenvalue efficiently?

I have a method that uses a `noise_tol` parameter and searches for eigenvalues larger than this. Looking at cora, there should ~78 zero eigenvalues corresponding to the ~78 connected components and what we see is 78 eigen values on < ` 1e-4`, and the next eigenvalue is ~`1e-2` so there's quite a clear gap between.

From what I've noticed, the auto calculated scaling factors  perform really poorly and regardless of the graph seem to be < 1. I found the code from the paper and they use manual scaling factors of `[19, 20, 21]`. I'm thinking of dropping the scaling factor code because:

1. It doesn't seem to work well in practice.
2. The authors use much larger scaling factors in their published results.
3. With larger scaling factors the algorithm is much more efficient.

What do you think?

Reply:
> I now a fast method of searching the eigenvalues; 0.25s to recover 80% of the filter eigenvalue norm on Cora with min_scale=4.

Nice!

> I have a method that uses a noise_tol parameter and searches for eigenvalues larger than this. Looking at cora, there should ~78 zero eigenvalues corresponding to the ~78 connected components and what we see is 78 eigen values on < 1e-4, and the next eigenvalue is ~1e-2 so there's quite a clear gap between.

I guess I'll wait to see how this works, but how easy is it for a user to pass in a correct value?

> What do you think?
 
Sounds great, although please document the difference to the paper somewhere (and especially why we differ).

Reply:
> I guess I'll wait to see how this works, but how easy is it for a user to pass in a correct value?
> 

I've removed the scale calculation now so we no longer need to calculate the smallest non-zero eigenvalue.